main.cpp

```diff
@@ -318,460 +267,310 @@ struct HeapA
  Wait for my code review.
  */
 
-
-struct FloatType
+template<typename T>
+struct Numeric
 {
-    explicit FloatType(float f) : value(new float(f)) {}
-    ~FloatType()
-    {
-        delete value;
-    } 
-
-    FloatType& operator+=(float lhs);
-    FloatType& operator-=(float lhs);
-    FloatType& operator*=(float lhs);
-    FloatType& operator/=(float lhs);
+    using Type = T;
 
-    FloatType& pow(FloatType&);
-    FloatType& pow(DoubleType&);
-    FloatType& pow(IntType&);
-    FloatType& pow(float);
+    Numeric(Type num) : value(std::make_unique<Type>(num)) {}
+    ~Numeric() {}
 
-    FloatType& apply(std::function<FloatType&(float&)> func);
-    FloatType& apply(void(*func)(float&));
-
-    operator float() const
+    Numeric& operator+=(Type rhs) 
     {
-        return *value;
+        *value += rhs;
+        return *this;
     }
     
-
-private:
-    float* value;
-    FloatType& powInternal(const float);
-};
-
-struct DoubleType
-{
-    explicit DoubleType(double d) : value(new double(d)) {}
-    ~DoubleType()
+    
+    Numeric& operator-=(Type rhs)
+    {       
+        *value -= rhs;
+        return *this;  
+    }
+    
+    
+    Numeric& operator*=(Type rhs)
     {
-        delete value;
-    } 
-
-    DoubleType& operator+=(double lhs);
-    DoubleType& operator-=(double lhs);
-    DoubleType& operator*=(double lhs);
-    DoubleType& operator/=(double lhs);
-
+        *value *= rhs;
+        return *this;
+    }
+    
+    
+    template<typename P>
+    Numeric& operator/=(P rhs)
+    {
+        if constexpr(std::is_same<int, Type>::value)
+        {
+            if constexpr(std::is_same<int, P>::value)
+            {
+                if(rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+                    return *this;
+                }
+            }
+            else if (static_cast<float>(rhs) < std::numeric_limits<float>::epsilon())
```

Comment:
you have the wrong types here, and you shouldn't be casting either
- [ ] revise